server/container/util-java8/src/test/java/org/apache/james/util/CompletableFutureUtilTest.java

```diff
@@ -95,4 +96,45 @@ public void allOfShouldWorkOnVeryLargeStream() {
                 .map(CompletableFuture::completedFuture))
             .join();
     }
+
+    @Test
+    public void mapOnAllShouldMapOnStreamInsideACompletableFuturOfStream() {
+        CompletableFuture<Stream<Integer>> futurOfInteger = CompletableFuture.completedFuture(Stream.of(1,2,3));
+
+        assertThat(
+            CompletableFutureUtil.map(futurOfInteger, integer ->
+                integer * 2)
+                .join()
+                .collect(Guavate.toImmutableList()))
+            .containsExactly(2,4,6);
+    }
+
+    @Test
+    public void flatMapOnAllShouldMapOnStreamInsideACompletableFuturOfStreamAndTransformTheResultingStreamOfCompletableFutureIntoACompletableOfStreamAndFlapIt() {
+        CompletableFuture<Stream<Integer>> futurOfInteger = CompletableFuture.completedFuture(Stream.of(1,2,3));
+
+        assertThat(
+            CompletableFutureUtil.thenComposeOnAll(futurOfInteger, integer ->
+                CompletableFuture.completedFuture(integer * 2))
+                .join()
+                .collect(Guavate.toImmutableList()))
+            .containsExactly(2,4,6);
+    }
+
+    @Test
+    public void keepValueShouldCompleteWhenTheGivenCompletableFutureEnd() {
```

Comment:
Otherwise we could use : http://www.ontestautomation.com/an-introduction-to-property-based-testing-with-junit-quickcheck/